### PR TITLE
[USETUP] Speed up DoFileCopy +24%

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
Save our lifetime.
JIRA issue: [CORE-18838](https://jira.reactos.org/browse/CORE-18838)

## Proposed changes

- Make `STRING_COPYING` cached.

BEFORE:
```txt
(base/setup/usetup/usetup.c:159) StopWatch: DoFileCopy: 857125632
```
AFTER:
```txt
(base/setup/usetup/usetup.c:159) StopWatch: DoFileCopy: 650791680
```
+24% speed up in `DoFileCopy`.

## TODO

- [x] Do tests.

### CODE REVIEW

File: base/setup/usetup/usetup.c [hunk 1]
```diff
@@ -143,6 +143,23 @@ DoWatchDestFileName(LPCWSTR FileName)
 
 /* FUNCTIONS ****************************************************************/
 
+VOID StopWatch(LPCSTR Name OPTIONAL)
```
Comment (HBelusca) [inline base/setup/usetup/usetup.c:146]: Debug code to be removed for final PR version.

**Comment by HBelusca** [comment]
Of course there will be a "speedup". But is it actually visible? (without the need to add a perf function?)

File: base/setup/usetup/usetup.c (hunk 1 above)
Comment (katahiromz) [reply base/setup/usetup/usetup.c]: Deleted.

**Review by JoachimHenze — APPROVED** [review]
Code-wise I have to admit that I do like the usetup.c changes most. :-)
But speed is speed. Approved for that reason.

**Comment by katahiromz** [comment]
> But is it actually visible?

It became slightly smoothy.

**Review by HBelusca — CHANGES_REQUESTED** [review]
I think that a much simpler (and less invasive) way to do the speed up, would be to just cache the (pointer to) strings in the FileCopyCallback function itself:
```c
static PCSTR s_pszCopying = NULL;
```
(the delete, move or rename do not need the caching at all because they are very seldomly used; but if they are ever used later -- for update scenarii -- the same could be done),
then do the usual:
```c
if (!s_pszCopying)
    s_pszCopying = MUIGetString(STRING_COPYING);
```
Then, in the copy step, replace the call with
```c
CONSOLE_SetStatusText(s_pszCopying, DstFileName);
```

**Comment by katahiromz** [comment]
OK.

File: base/setup/usetup/usetup.c [hunk 2]
```diff
@@ -3658,8 +3660,9 @@ FileCopyCallback(PVOID Context,
                 if (DstFileName) ++DstFileName;
                 else DstFileName = FilePathInfo->Target;
 
-                CONSOLE_SetStatusText(MUIGetString(STRING_COPYING),
-                                      DstFileName);
+                if (s_pszCopying == NULL)
```
Comment (HBelusca) [inline base/setup/usetup/usetup.c:3663]:
```suggestion
                if (!s_pszCopying)
```

**Review by HBelusca — APPROVED** [review]

**Comment by JoachimHenze** [comment]
https://build.reactos.org/#/builders/3/builds/20406 0.4.15-dev-5718-gf1de615 f1de615 VBox [SYSREG] Stage took: 33.335868 seconds
https://build.reactos.org/#/builders/3/builds/20408 0.4.15-dev-5719-gacd3148 acd3148 VBox [SYSREG] Stage took: 33.146574 seconds
https://build.reactos.org/#/builders/3/builds/20416 0.4.15-dev-5726-g13cf4c5 13cf4c5 VBox [SYSREG] Stage took: 33.338083 seconds "Katayamas 24% speedup PR5089"

https://build.reactos.org/#/builders/6/builds/32937 0.4.15-dev-5718-gf1de615 f1de615 KVM [SYSREG] Stage took: 47.870212 seconds
https://build.reactos.org/#/builders/6/builds/32939 0.4.15-dev-5719-gacd3148 acd3148 KVM [SYSREG] Stage took: 42.906221 seconds
https://build.reactos.org/#/builders/6/builds/32940 0.4.15-dev-5720-g4c1e83d 4c1e83d KVM [SYSREG] Stage took: 47.949083 seconds
https://build.reactos.org/#/builders/6/builds/32941 0.4.15-dev-5721-gead9366 ead9366 KVM [SYSREG] Stage took: 48.053923 seconds
https://build.reactos.org/#/builders/6/builds/32947 0.4.15-dev-5726-g13cf4c5 13cf4c5 KVM [SYSREG] Stage took: 40.209613 seconds "Katayamas 24% speedup PR5089"

Based on just a single runs result, I'd say that KVM maybe is a bit faster, but VBox does not see any improvement.
More runs might make sense.

**Pull request merged**